Only evaluate particle velocities if needed

gassmoeller opened the pull request:
This is a change to the particle advection algorithm that I wanted to implement for a while. Each particle integrator only needs a subset of the available velocity solutions on the particle location for each integrator step. E.g. the forward euler integrator only needs the old solution to move particles, while RK2 and RK4 sometimes also need particle velocities at other timesteps (depending on their integrator step). This PR implements an interface for the particle integrators to tell the particle world which velocities are needed at the current integrator step. This allows the world to only compute the velocities that are needed. 

Because of all the other operations that are needed for the advection this optimization is not huge (it saves ~5% of the particle advection time), but it is also not a big change so I think it is worth it. All the relevant change happens in world.cc, and it is only a resorting of code lines so that I can make the operations dependent on the result of the `required_solution_vectors` function.

bangerth commented on include/aspect/particle/integrator/euler.h:
```diff
@@ -66,6 +66,20 @@ namespace aspect
+           * indicates if the particle integrator requires the solution vector
+           * at the new time (k+1).
+           *
+           * The forward euler integrator only requires the solution vector at the
+           * old time (k).
```
```suggestion
           * The forward Euler integrator only requires the solution vector at the
           * old time (k), and consequently returns `{false, true, false}`.
```

bangerth commented on source/particle/integrator/interface.cc:
```diff
@@ -68,6 +68,15 @@ namespace aspect
 
 
 
+      template <int dim>
+      std::array<bool, 3>
+      Interface<dim>::required_solution_vectors() const
+      {
+        return {{false, true, true}};
+      }
```
Why is this the right default? I would have expected that you err on the safe side and return `false, false, false`. But perhaps it would make sense to just not implement this function at all (i.e., mark it as `=0` in the `.h` file)?

gassmoeller commented:
Thanks for the comments. I think both defaults (=0 or {false, true, true}) are reasonable. I initially chose the last one, because it would be fully compatible with any existing plugins (it models the behavior before this PR). But you are right, requiring plugins to provide the function forces them to think about what is actually needed. In practice I dont think the default matters that much, because I highly doubt anyone has written their own particle integrator. I am fine with using =0.

I addressed all of your comments.
